walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -28,44 +42,82 @@
   private static final TimeRange WHOLE_DAY = TimeRange.WHOLE_DAY;
   private static final Comparator<TimeRange> ORDER_BY_START = TimeRange.ORDER_BY_START;
 
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+  public Collection<TimeRange> query(
+      Collection<Event> events, MeetingRequest request) {
 
     // Array of queries representing times to meet.
-    List<TimeRange> query = new ArrayList<>();
+    List<TimeRange> optionalTimeRanges;
+
+    // Array of queries representing times to meet.
+    List<TimeRange> mandatoryTimeRanges;
+
+    // Array of queries representing times to meet.
+    List<TimeRange> allTimeRanges;
 
     int desiredDuration = (int) request.getDuration();
 
     // Collection of string representing who is to attend the requested meeting.
-    Collection<String> requestAttendees = request.getAttendees();
+    Collection<String> mandatoryAttendees = new ArrayList<>(request.getAttendees());
+
+    // Collection of strings representing who are optional to the meeting.
+    Collection<String> optionalAttendees = new ArrayList<>(request.getOptionalAttendees());
+
+    // Collection of Strings representing all attendees ( Mandatory and Optional )
+    Collection<String> allAttendees =
+        combineMandatoryAndOptional(mandatoryAttendees, optionalAttendees);
 
     // Check if duration of event is longer than a whole day and if so return empty.
     if ((int) request.getDuration() > WHOLE_DAY.duration()) {
-      return query;
+      return Collections.emptyList();
     }
 
-    // Check if there are any events happening at all if so return whole day query.
-    if (events.size() == 0) {
-      query.add(TimeRange.WHOLE_DAY);
-      return query;
+    // If there are no mandatory attendees , return lists of the optional attendees time ranges.
+    if (mandatoryAttendees.isEmpty()) {
+        return getTimeRanges(events, optionalAttendees, desiredDuration);
     }
 
+    // If there are no optional attendees , return lists of the mandatory attendees time ranges.
+    if (optionalAttendees.isEmpty()) {
+        return getTimeRanges(events, mandatoryAttendees, desiredDuration);
+    }
+
+    allTimeRanges =
+        (ArrayList<TimeRange>) getTimeRanges(events, allAttendees, desiredDuration);
+    if (allTimeRanges.isEmpty()) {
+        return getTimeRanges(events, mandatoryAttendees, desiredDuration);
+    }
+    return allTimeRanges;
+  }
+
+ public List<TimeRange> getTimeRanges(
+      Collection<Event> events, Collection<String> attendees, int desiredDuration) {
+
+    // Array of queries representing times to meet.
+    List<TimeRange> timeRanges = new ArrayList<>();
+
+    int nextTime = START_OF_DAY;
```

Comment:
Move this closer to the for loop, because if the function exits in line 109, you don't need this variable anymore.

Reply:
moved down !